File: ios/Sources/SduiCore/Rendering/SectionErrorBoundary.swift
import SwiftUI
import os

private let logger = Logger(subsystem: "com.nba.sdui", category: "SectionErrorBoundary")

/// Per-section error boundary. Mirrors Android's
/// [`SectionErrorBoundary`](../../../../../android/sdui-core/src/main/java/com/nba/sdui/core/renderer/SectionErrorBoundary.kt).
///
/// SwiftUI (like Compose) can't wrap view bodies with `try/catch`, so the
/// boundary runs pre-validation on the section's `data` before rendering
/// and also respects explicit error propagation via ``reportError(_:)``.
/// Supports:
/// - `sectionStates.error.hideOnError` → collapses to `EmptyView`.
/// - `sectionStates.error.message`     → server-provided human message.
/// - `sectionStates.error.retryAction` → dispatched via `onAction`.
/// - Retry budget (default 5) to prevent infinite retry loops.
struct SectionErrorBoundary<Content: View>: View {
    let sectionID: String
    let sectionType: String
    let sectionStates: SectionStates?
    let data: [String: Any]?
    let onAction: (Action) -> Void
    let maxRetries: Int
    @ViewBuilder let content: () -> Content

    @State private var retryCount: Int = 0
    @State private var explicitError: String?

    init(
        sectionID: String,
        sectionType: String,
        sectionStates: SectionStates?,
        data: [String: Any]?,
        onAction: @escaping (Action) -> Void,
        maxRetries: Int = 5,
        @ViewBuilder content: @escaping () -> Content
    ) {
        self.sectionID = sectionID
        self.sectionType = sectionType
        self.sectionStates = sectionStates
        self.data = data
        self.onAction = onAction
        self.maxRetries = maxRetries
        self.content = content
    }

    var body: some View {
        let validationError = SectionErrorBoundary.validate(
            sectionID: sectionID,
            sectionType: sectionType,
            data: data
        )
        let activeError = explicitError ?? validationError

        if let message = activeError {
            let _ = logger.error("section render failed id=\(self.sectionID, privacy: .public) type=\(self.sectionType, privacy: .public) error=\(message, privacy: .public)")

            let errorConfig = sectionStates?.error
            if errorConfig?.hideOnError == true {
                EmptyView()
            } else {
                SectionErrorCard(
                    message: errorConfig?.message ?? message,
                    retryAction: errorConfig?.retryAction,
                    canRetry: retryCount < maxRetries,
                    onRetry: {
                        retryCount += 1
                        explicitError = nil
                    },
                    onAction: onAction
                )
            }
        } else {
            content()
        }
    }

    /// Pre-validation rules shared with Android's validateSection. Returns
    /// a human-readable message when rendering should be suppressed.
    static func validate(sectionID: String, sectionType: String, data: [String: Any]?) -> String? {
        if sectionType == OverlayType.atomicComposite.rawValue, data?["ui"] == nil {
            return "AtomicComposite section \(sectionID) has no ui element"
        }
        return nil
    }
}

// Extension access so SectionLayout can nudge the boundary into error
// state when e.g. a decode fails outside the view body.
extension SectionErrorBoundary {
    func reportError(_ message: String) {
        explicitError = message
    }
}

private struct SectionErrorCard: View {
    let message: String
    let retryAction: Action?
    let canRetry: Bool
    let onRetry: () -> Void
    let onAction: (Action) -> Void

    var body: some View {
        VStack(spacing: 12) {
            Image(systemName: "exclamationmark.triangle.fill")
                .font(.title)
                .foregroundStyle(.orange)
            Text(message)
                .font(.body)
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)
            if canRetry, let retryAction {
                Button {
                    onAction(retryAction)
                    onRetry()
                } label: {
                    Text("Try Again")
                }
                .buttonStyle(.bordered)
            }
        }
        .padding(24)
        .frame(maxWidth: .infinity)
        .background(
            RoundedRectangle(cornerRadius: 12, style: .continuous)
                .fill(Color.red.opacity(0.08))
        )
        .padding(.vertical, 4)
    }
}

File: android/sdui-core/src/main/java/com/nba/sdui/core/renderer/SectionErrorBoundary.kt
package com.nba.sdui.core.renderer

import android.util.Log
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.nba.sdui.core.models.generated.Action
import com.nba.sdui.core.models.generated.Data
import com.nba.sdui.core.models.generated.SectionStates
import com.nba.sdui.core.renderer.adapters.toSduiAction
import com.nba.sdui.core.state.SduiAction

/**
 * Per-section error boundary — pre-validation + error state tracking.
 *
 * Since Compose does not allow try/catch around @Composable invocations,
 * this boundary uses two strategies:
 *
 * 1. **Pre-validation**: Validates section data before rendering. Catches
 *    data problems (nulls, type mismatches, missing fields) which are ~95%
 *    of real-world failures.
 *
 * 2. **Error state propagation**: Sections can report errors via the
 *    onSectionError callback, which triggers the error card display.
 *    True recomposition crashes are captured by app-level crash telemetry
 *    (New Relic etc.), not per-section isolation.
 *
 * Supports:
 * - hideOnError (collapse section entirely)
 * - retryAction from sectionStates.error
 * - Configurable retry budget (default 5)
 * - §12-compliant error logging
 */
private const val TAG = "SectionErrorBoundary"

/**
 * Validates section data before rendering. Returns an error message if
 * the section should not be rendered, or null if validation passes.
 */
fun validateSection(
    sectionId: String,
    sectionType: String,
    data: Data?
): String? {
    // AtomicComposite requires a ui element in data
    if (sectionType == "AtomicComposite" && data?.ui == null) {
        return "AtomicComposite section $sectionId has no ui element"
    }
    return null
}

@Composable
fun SectionErrorBoundary(
    sectionId: String,
    sectionType: String,
    sectionStates: SectionStates?,
    data: Data?,
    onAction: (SduiAction) -> Unit,
    maxRetries: Int = 5,
    content: @Composable () -> Unit
) {
    var errorMessage by remember { mutableStateOf<String?>(null) }
    var retryCount by remember { mutableIntStateOf(0) }

    // Pre-validation: check section data before attempting to render
    val validationError = remember(sectionId, sectionType, data) {
        validateSection(sectionId, sectionType, data)
    }

    val activeError = errorMessage ?: validationError

    if (activeError != null) {
        Log.e(TAG, "Section render failed: id=$sectionId type=$sectionType error=$activeError")

        val errorConfig = sectionStates?.error

        // hideOnError: collapse section entirely
        if (errorConfig?.hideOnError == true) {
            return
        }

        SectionErrorCard(
            message = errorConfig?.message ?: activeError,
            retryAction = errorConfig?.retryAction,
            canRetry = retryCount < maxRetries,
            onRetry = {
                retryCount++
                errorMessage = null
            },
            onAction = onAction
        )
    } else {
        content()
    }
}

@Composable
private fun SectionErrorCard(
    message: String,
    retryAction: Action?,
    canRetry: Boolean,
    onRetry: () -> Unit,
    onAction: (SduiAction) -> Unit
) {
    Surface(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 0.dp, vertical = 4.dp),
        color = MaterialTheme.colorScheme.errorContainer.copy(alpha = 0.3f),
        shape = MaterialTheme.shapes.medium
    ) {
        Column(
            modifier = Modifier.padding(24.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Text(
                text = "⚠",
                style = MaterialTheme.typography.titleLarge
            )
            Spacer(modifier = Modifier.height(8.dp))
            Text(
                text = message,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.error
            )
            if (canRetry && retryAction != null) {
                Spacer(modifier = Modifier.height(16.dp))
                OutlinedButton(onClick = {
                    retryAction?.let { onAction(it.toSduiAction()) }
                    onRetry()
                }) {
                    Text("Try Again")
                }
            }
        }
    }
}
